clang/lib/Analysis/FlowSensitive/Models/ReverseNullModel.cpp

```diff
@@ -0,0 +1,723 @@
+  return CFGMatchSwitchBuilder<Environment>()
+      .CaseOfCFGStmt<Stmt>(derefMatcher(), matchDereferenceExpr)
+      .CaseOfCFGStmt<Stmt>(arrowMatcher(), matchDereferenceExpr)
+      .CaseOfCFGStmt<Expr>(nullptrMatcher(), matchNullptrExpr)
+      .CaseOfCFGStmt<Expr>(addressofMatcher(), matchAddressofExpr)
+      .CaseOfCFGStmt<Expr>(functionCallMatcher(), matchFunctionCallExpr)
+      .CaseOfCFGStmt<Expr>(anyPointerMatcher(), matchFunctionCallExpr)
+      .Build();
+}
+
+auto buildBranchTransferMatchSwitch() {
+  return ASTMatchSwitchBuilder<Stmt, ReverseNullModel::TransferArgs>()
+      .CaseOf<CastExpr>(castExprMatcher(), matchCastExpr)
+      .Build();
+}
+
+auto buildDiagnoseMatchSwitch() {
+  return CFGMatchSwitchBuilder<const Environment, ReverseNullDiagnoser::ResultType>()
+      .CaseOfCFGStmt<CastExpr>(castExprMatcher(), diagnoseCastExpr)
+      .Build();
+}
+
+} // namespace
+
+ReverseNullModel::ReverseNullModel(ASTContext &Context)
+    : DataflowAnalysis<ReverseNullModel, NoopLattice>(Context),
+      TransferMatchSwitch(buildTransferMatchSwitch()),
+      BranchTransferMatchSwitch(buildBranchTransferMatchSwitch()) {
+}
+
+ast_matchers::StatementMatcher ReverseNullModel::ptrValueMatcher() {
+  return ptrToVar();
+}
+
+void ReverseNullModel::initializeFunctionParameters(const ControlFlowContext &CFCtx, Environment &Env) {
+  const FunctionDecl &FD = cast<FunctionDecl>(CFCtx.getDecl());
+
+  for (const auto *Param : FD.parameters()) {
+    if (!Param || !Param->getType()->isAnyPointerType())
+      continue;
+
+    StorageLocation *SL = Env.getStorageLocation(*Param);
+    if (!SL) {
+      SL = &Env.createStorageLocation(*Param);
+      Env.setStorageLocation(*Param, *SL);
+    }
+
+    auto *RootValue = Env.getValue(*Param);
+    if (!RootValue) {
+      RootValue = Env.createValue(Param->getType());
+      if (!RootValue)
+        continue;
+      Env.setValue(*SL, *RootValue);
+    }
+
+    initializeRootValue(*RootValue, Env);
+  }
+}
+
+void ReverseNullModel::transfer(const CFGElement &E, NoopLattice &State,
+                                Environment &Env) {
+  TransferMatchSwitch(E, getASTContext(), Env);
+}
+
+void ReverseNullModel::transferBranch(bool Branch, const Stmt *E, NoopLattice &State,
+                                      Environment &Env) {
+  if (!E)
+    return;
+
+  TransferArgs Args = { Branch, Env };
+  BranchTransferMatchSwitch(*E, getASTContext(), Args);
+}
+
+// FIXME: match any true into true, and only falses into false
+bool ReverseNullModel::merge(QualType Type, const Value &Val1,
+                             const Environment &Env1, const Value &Val2,
+                             const Environment &Env2, Value &MergedVal,
+                             Environment &MergedEnv) {
+    if (!Type->isAnyPointerType())
+      return false; // FIXME: Maybe requires true?
+
+    // Alternative attempt: is any branch is true ie. can be a pointer, merging it is also a pointer
```

Comment:
Can be a pointer, or can be a **null** pointer?
Does "can be a pointer" mean that it is not null?